File: src/algorithm/sorting/selection-sort/selection_sort.py
#! /usr/local/bin/python3


class SelectionSort:
    def sort(self, original_list):
        if original_list is None:
            return []
        arr = original_list.copy()
        if len(arr) <= 1:
            return arr

        left_list = [arr.pop()]

        for i, item in enumerate(arr):
            if item < left_list[0]:
                left_list[0], arr[i] = arr[i], left_list[0]

        right_list = self.sort(arr)

        return left_list + right_list

    def test(self):
        import time

        start = time.process_time()
        result = self.sort([34, 6, 44, 5, 343, 6, 9, 6, 7, 7, 34, 454, 9])
        end = time.process_time()

        print(result, end - start)

if __name__ == '__main__':
    SelectionSort().test()

File: src/algorithm/sorting/selection-sort/selectionSort.js
#! /usr/local/bin/node


class SelectionSort {
    sort(originalList) {
        let list = [...originalList];

        if (list.length <= 1) {
            return list;
        }

        let leftList = [list.shift()];

        while (list.length > 0) {
            for (let i = 0; i < list.length; i++) {
                if (list[i] < leftList[leftList.length - 1]) {
                    [leftList[leftList.length - 1], list[i]] = [list[i], leftList[leftList.length - 1]];
                }
            }

            leftList.push(list.shift());
        }

        return leftList;
    }

    isSorted(array=[]) {
        if (array.length <= 1) {
            return true;
        }

        for (let i = 1; i < array.length; i++) {
            if (array[i] < array[i - 1]) {
                return false;
            }
        }

        return true;
    }

    test() {
        const array = new Array(40000).fill(undefined).map(item => Math.random() * 10000);
        console.time('sort');
        let result = this.sort(array);
        console.timeEnd('sort');
        let sorted = this.isSorted(result);
        console.log('sorted:', sorted);
    }
}

new SelectionSort().test();
